### ISSUE TITLE
Handle closed websocket connections

### ISSUE THREAD

**Issue by seamusallen0905** [jacoduplessis/luno_streams#2]
I have been trying to use your example code to stream data from Luno.

All works until anything from a a few minutes to a few hours when I get the fault coming through : code = 1006 (connection closed abnormally [internal]), no reason

```
Example usage
import asyncio
from luno_streams import Updater

def print_it(consolidated_order_book, trades):
    print(consolidated_order_book, trades)

updater = Updater(
    pair_code='XBTZAR',
    api_key='123',
    api_secret='456',
    hooks=[print_it],
)

loop = asyncio.get_event_loop()
try:
    loop.run_until_complete(updater.run())
except Exception as e:
    print('Problem Occurred in loop.run')
    print(e)
    pass
```

Obviously the api key and secret are my credentials.

Any help would be appreciated here

Thanks

**Comment by jacoduplessis** [comment]
@seamusallen0905 can you post the stack trace here?

**Comment by seamusallen0905** [comment]
@jacoduplessis Thanks for the reply, Will run the code again till it fails, but this time without the try: except: and post the data - Thanks

**Comment by seamusallen0905** [comment]
@jacoduplessis : Here we go:

```
Traceback (most recent call last):
  File "C:\Python37\lib\site-packages\websockets\protocol.py", line 795, in transfer_data
    message = await self.read_message()
  File "C:\Python37\lib\site-packages\websockets\protocol.py", line 863, in read_message
    frame = await self.read_data_frame(max_size=self.max_size)
  File "C:\Python37\lib\site-packages\websockets\protocol.py", line 938, in read_data_frame
    frame = await self.read_frame(max_size)
  File "C:\Python37\lib\site-packages\websockets\protocol.py", line 1018, in read_frame
    extensions=self.extensions,
  File "C:\Python37\lib\site-packages\websockets\framing.py", line 121, in read
    data = await reader(2)
  File "C:\Python37\lib\asyncio\streams.py", line 677, in readexactly
    raise IncompleteReadError(incomplete, n)
asyncio.streams.IncompleteReadError: 0 bytes read on a total of 2 expected bytes

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "updater.py", line 17, in <module>
    loop.run_until_complete(updater.run())
  File "C:\Python37\lib\asyncio\base_events.py", line 579, in run_until_complete
    return future.result()
  File "C:\Python37\progs\luno_streams\luno_streams\updater.py", line 67, in run
    async for message in self.websocket:
  File "C:\Python37\lib\site-packages\websockets\protocol.py", line 414, in __aiter__
    yield await self.recv()
  File "C:\Python37\lib\site-packages\websockets\protocol.py", line 484, in recv
    await self.ensure_open()
  File "C:\Python37\lib\site-packages\websockets\protocol.py", line 771, in ensure_open
    raise self.connection_closed_exc()
websockets.exceptions.ConnectionClosedError: code = 1006 (connection closed abnormally [internal]), no reason
```

I Hope this helps - Thanks

By the way, it ran for about 2 hours

**Comment by jacoduplessis** [comment]
Seems like the websocket connection is simply lost. Are you running this on a server or home internet?

Either way, I think it would be a good idea to add some logic to handle closed connections by automatically reconnecting. Thanks for raising this issue.

**Comment by seamusallen0905** [comment]
I fully agree that this could be the problem, just not sure where to restart it from scratch, but thats a challenge I am willing to accept.......
Thanks for the help Jaco
Cheers

**Comment by jacoduplessis** [comment]
Why don't you give something like this a try:

```python
from websockets.exceptions import ConnectionClosedError

while True:
    try:
        loop.run_until_complete(updater.run())
    except ConnectionClosedError:
        continue
```

**Comment by seamusallen0905** [comment]
I have tried something similar and if it works I shall share with you the result, but I'm not hopeful just yet that it will work.  If not I will try your method, which I may add, am very appreciated of.
Can you tell me how this logger works? Been trying to figure it out from the logging module on https://docs.python.org/3.5/library/logging.html. Where does it actually output to?
Sorry, I am a noob, but learning every day.
Thanks again for the help here Jaco

**Comment by seamusallen0905** [comment]
@jacoduplessis 
My method certainly did not work at all, lol. So tried yours, seems a lot better, but still faults. Seems like it does not fault much if Luno is really busy, but when the trades slow down is when the issue maybe happens - I don't know. It could be that something is timing out maybe
Here is the latest trace, when is better than the first trace I presented at the start.
```
Traceback (most recent call last):
  File "updater.py", line 20, in <module>
    loop.run_until_complete(updater.run())
  File "C:\Python37\lib\asyncio\base_events.py", line 579, in run_until_complete
    return future.result()
  File "C:\Python37\progs\luno_streams\luno_streams\updater.py", line 66, in run
    await self.connect()
  File "C:\Python37\progs\luno_streams\luno_streams\updater.py", line 40, in connect
    await self.websocket.ws_client.close()
AttributeError: 'WebSocketClientProtocol' object has no attribute 'ws_client'

```
Regards

**Comment by seamusallen0905** [comment]
@jacoduplessis I have made a change on this line


```
        if self.websocket is not None:  # reconnecting
            logger.info(f'[{self.pair_code}] Closing existing connection...')
            await self.websocket.close() # await self.websocket.ws_client.close()
```

Its been running the last 3 days now with one stop, but that had to do with it not seeing a 'buy', but just did this to bypass the fault for the time being:


```
        initial = await self.websocket.recv()
        initial_data = json.loads(initial)
        try:
            self.sequence = int(initial_data['sequence'])
            self.asks = {x['id']: [Decimal(x['price']), Decimal(x['volume'])] for x in initial_data['asks']}
            self.bids = {x['id']: [Decimal(x['price']), Decimal(x['volume'])] for x in initial_data['bids']}
            logger.info(f'[{self.pair_code}] Initial state received.')
        except Exception as e:
            print('Error processing the received data')
            pass
```

Tell me if you agree, might be a small bug possibly?

I would change the software on github, but no idea how to do this so will leave this up to you, the expert

Thanks a mil

Sean

**Comment by jacoduplessis** [comment]
Thanks Sean.

The reference to `ws_client` was a bug that has now been fixed in 1e6adde8b6b2be4f3d0dbace705787a0d6a0be87.

**Comment by jacoduplessis** [comment]
> Can you tell me how this logger works? Been trying to figure it out from the logging module on https://docs.python.org/3.5/library/logging.html. Where does it actually output to?

You have to attach a handler to see the output. You can add a logging.StreamHandler to output to your terminal. The Python docs have some good examples.
